File: dirloop.go
package main

import (
	"bytes"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"path"
	"path/filepath"
)

// Arguments holds the command line arguments
type Arguments struct {
	Dir              string
	DirPattern       string
	DryRun           bool
	CommandName      string
	CommandArguments []string
}

// CreateCommand creates an exec.Cmd instance based on the given arguments.
func (a Arguments) CreateCommand() *exec.Cmd {
	return exec.Command(a.CommandName, a.CommandArguments...)
}

// IsMatchingDir checks if the given argument represents a directory
// and that directory's name matches the directory pattern.
func (a Arguments) IsMatchingDir(fileInfo os.FileInfo) (bool, error) {
	if !fileInfo.IsDir() {
		return false, nil
	}

	if a.DirPattern == "" {
		return true, nil
	}

	return path.Match(a.DirPattern, fileInfo.Name())
}

// Parses the command line arguments
func parseArguments() Arguments {
	var arguments Arguments
	flag.StringVar(&arguments.Dir, "dir", ".", "The root directory containing other directories")
	flag.StringVar(&arguments.DirPattern, "dir-pattern", "", "An optional wildcard pattern to select only some directories")
	flag.BoolVar(&arguments.DryRun, "dry-run", false, "Don't actually run any command")

	flag.Parse()

	remainingArguments := flag.Args()
	if len(remainingArguments) <= 0 {
		fmt.Println("Please specify the command that should be run")
		os.Exit(1)
	}

	arguments.CommandName = remainingArguments[0]
	arguments.CommandArguments = remainingArguments[1:]
	return arguments
}

func runCommand(arguments Arguments, dirName string) error {
	fullDir := filepath.Join(arguments.Dir, dirName)
	if arguments.DryRun {
		fmt.Printf("Would have run command in %s", fullDir)
		fmt.Println()
		return nil
	}

	fmt.Printf("Running command in %s", fullDir)
	fmt.Println()

	var out bytes.Buffer
	cmd := arguments.CreateCommand()
	cmd.Stdout = &out
	cmd.Dir = fullDir
	err := cmd.Run()
	if err != nil {
		return err
	}

	fmt.Printf(out.String())
	fmt.Println()
	return nil
}

func main() {
	arguments := parseArguments()

	files, err := ioutil.ReadDir(arguments.Dir)
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}

	for _, file := range files {
		isMatchingDir, err := arguments.IsMatchingDir(file)
		if err != nil {
			log.Fatal(err)
		} else if isMatchingDir {
			err := runCommand(arguments, file.Name())
			if err != nil {
				fmt.Println(err)
			}
		}
	}
}

File: dirloop.js
const { Command } = require('commander');
const fs = require('fs');
require('process');
const { spawnSync } = require('child_process');
const path = require('path');
const chalk = require('chalk');
const { name, version } = require('./package.json');

/**
 * @type {{dir: string, dirPrefix: string, args: string[], dryRun: boolean}}
 */
let args = null;

function log(message) {
  console.log(message);
}

function logVerbose(message) {
  if (args.verbose) {
    log(chalk.blue(message));
  }
}

function logError(message) {
  log(chalk.red(message));
}

/**
 * Checks if the given argument represents a directory
 * and that directory's name matches the directory pattern.
 */
function isMatchingDir(file) {
  if (!file.isDirectory()) {
    return false;
  }

  if (args.dirPrefix) {
    return file.name.startsWith(args.dirPrefix);
  }

  return true;
}

/**
 * Parses the command line arguments.
 */
function parseArguments() {
  const commander = new Command();

  commander
    .name(name)
    .version(version)
    .option('--dir <dir>', 'The root directory containing other directories')
    .option(
      '--dir-prefix <prefix>',
      'An optional prefix to select only some directories',
    )
    .option('--dry-run', "Don't actually run any command")
    .option('-- <command> [options...]', 'The command to run')
    .option('-v, --verbose', 'Verbose output')
    .parse(process.argv);

  const defaults = {
    dir: '.',
    dirPrefix: '',
    dryRun: false,
  };

  Object.keys(defaults).forEach((key) => {
    commander[key] = commander[key] || defaults[key];
  });

  if (!commander.args || !commander.args.length) {
    logError('Command to run was not specified');
    commander.outputHelp();
    process.exit(1);
  }

  return commander;
}

function runCommand(file) {
  const absDir = path.resolve(args.dir, file.name);
  if (args.dryRun) {
    log(`Would have run command ${args.args.join(' ')} in ${absDir}`);
    return;
  }

  logVerbose(`Running command in ${absDir}`);

  const result = spawnSync(
    args.args[0],
    args.args.slice(1),
    {
      cwd: absDir,
      stdio: 'inherit',
    },
  );

  const { status } = result;

  if (status) {
    logError(chalk.red(`Command returned exit code ${status}`));
  }
}

function main() {
  args = parseArguments();

  const files = fs.readdirSync(args.dir, { withFileTypes: true });
  files.forEach((file) => {
    const matches = isMatchingDir(file);
    if (matches) {
      runCommand(file);
    }
  });
}

main();
